source/server/http_test_server_filter.cc

```diff
@@ -13,60 +13,52 @@ namespace Nighthawk {
 namespace Server {
 
 HttpTestServerDecoderFilterConfig::HttpTestServerDecoderFilterConfig(
-    nighthawk::server::ResponseOptions proto_config)
-    : server_config_(std::move(proto_config)) {}
+    const nighthawk::server::ResponseOptions& proto_config)
+    : FilterConfigurationBase(proto_config, "test-server") {}
 
 HttpTestServerDecoderFilter::HttpTestServerDecoderFilter(
     HttpTestServerDecoderFilterConfigSharedPtr config)
     : config_(std::move(config)) {}
 
 void HttpTestServerDecoderFilter::onDestroy() {}
 
-void HttpTestServerDecoderFilter::sendReply() {
-  if (!json_merge_error_) {
-    std::string response_body(base_config_.response_body_size(), 'a');
-    if (request_headers_dump_.has_value()) {
-      response_body += *request_headers_dump_;
-    }
-    decoder_callbacks_->sendLocalReply(
-        static_cast<Envoy::Http::Code>(200), response_body,
-        [this](Envoy::Http::ResponseHeaderMap& direct_response_headers) {
-          Configuration::applyConfigToResponseHeaders(direct_response_headers, base_config_);
-        },
-        absl::nullopt, "");
-  } else {
-    decoder_callbacks_->sendLocalReply(
-        static_cast<Envoy::Http::Code>(500),
-        fmt::format("test-server didn't understand the request: {}", error_message_), nullptr,
-        absl::nullopt, "");
+void HttpTestServerDecoderFilter::sendReply(const nighthawk::server::ResponseOptions& options) {
+  std::string response_body(options.response_body_size(), 'a');
+  if (request_headers_dump_.has_value()) {
+    response_body += *request_headers_dump_;
   }
+  decoder_callbacks_->sendLocalReply(
+      static_cast<Envoy::Http::Code>(200), response_body,
+      [options](Envoy::Http::ResponseHeaderMap& direct_response_headers) {
+        Configuration::applyConfigToResponseHeaders(direct_response_headers, options);
+      },
+      absl::nullopt, "");
 }
 
 Envoy::Http::FilterHeadersStatus
 HttpTestServerDecoderFilter::decodeHeaders(Envoy::Http::RequestHeaderMap& headers,
                                            bool end_stream) {
-  // TODO(oschaaf): Add functionality to clear fields
-  base_config_ = config_->server_config();
-  const auto* request_config_header = headers.get(TestServer::HeaderNames::get().TestServerConfig);
-  if (request_config_header) {
-    json_merge_error_ = !Configuration::mergeJsonConfig(
-        request_config_header->value().getStringView(), base_config_, error_message_);
-  }
-  if (base_config_.echo_request_headers()) {
-    std::stringstream headers_dump;
-    headers_dump << "\nRequest Headers:\n" << headers;
-    request_headers_dump_ = headers_dump.str();
-  }
+  config_->computeEffectiveConfiguration(headers);
   if (end_stream) {
-    sendReply();
+    if (!config_->maybeSendErrorReply(*decoder_callbacks_)) {
```

Comment:
Do you think that this would easier to read if `bool maybeSendErrorReply()` would be split up into `bool haveErrorReply()` and `void sendErrorReply()`?

Reply:
Probably. I'm not sure if I'm alone in feeling uncertain on which boolean value to expect from a maybe naming scheme though.

Reply:
What if we called it validateOrSendError.

Reply:
I was following up on this, I like the suggestion, but while doing that, I realised that the return value would have to be inverted. The `maybeXXX`  "naming scheme" usually returns true iff it did what it said it might do (`sendErrorReply()` in this case).  When we rename to `validateOrSendError()` the method should probably return true if the configuration was valid? I think this would still be a good change to make, but it would become a little intrusive as there's a bunch of places where this is used already. The good part is that when this PR gets merged, we are in a much better to position to make mechanical refactors like this across all extensions. Filed #558 to track following up.